File: src/python/core/application/use_cases/ask_question_use_case.py
from typing import List
from ..dtos.chat_request_dto import ChatRequestDto
from ..dtos.chat_result import ChatResult, ChatSource
from ..protocols.embeddings_protocol import EmbeddingsClient
from ..protocols.vector_store_protocol import VectorStore
from ..protocols.llm_protocol import LlmClient
from ...domain.configuration.vector_metadata_config import DEFAULT_VECTOR_METADATA_CONFIG
from ...infrastructure.shared.config_loader import get_config
from ...infrastructure.shared.prompt_loader import load_prompt
from ..services.query_parser import QueryParser
from ..services.candidate_aggregator import CandidateAggregator
from ..services.metadata_filter_builder import MetadataFilterBuilder
from ..services.candidate_ranker import CandidateRanker
from ...domain.configuration.ranking_weights import RankingWeights
from ...domain.entities.llm_response_schema import LlmResponseSchema, SelectedCandidate
import json

METADATA_CONFIG = DEFAULT_VECTOR_METADATA_CONFIG

DEFAULT_LIMIT = 6
IN_OPERATOR = "$in"
CONTEXT_SEPARATOR = "\n\n"
UNKNOWN_VALUE = "unknown"
DEFAULT_CONTENT_LIMIT = 200
CONTENT_SUFFIX = "..."
CHAT_SYSTEM_FILE = "chat_system.md"
CHAT_HUMAN_FILE = "chat_human.md"
ENGLISH_LEVEL_MAP = {"A1": 1, "A2": 2, "B1": 3, "B2": 4, "C1": 5, "C2": 6}

PREPARED_KEY = "prepared"
ENGLISH_LEVEL_NUM_MIN_KEY = "english_level_num_min"


class AskQuestionUseCase:
    def __init__(
        self, 
        embeddings_client: EmbeddingsClient, 
        vector_store: VectorStore, 
        llm_client: LlmClient
    ):
        self.embeddings_client = embeddings_client
        self.vector_store = vector_store
        self.llm_client = llm_client
        self.query_parser = QueryParser()
        self.candidate_aggregator = CandidateAggregator(METADATA_CONFIG.FIELD_CANDIDATE_ID)
        self.candidate_ranker = CandidateRanker(RankingWeights.default())
        self.filter_builder = MetadataFilterBuilder(
            candidate_id_field=METADATA_CONFIG.FIELD_CANDIDATE_ID,
            seniority_field=METADATA_CONFIG.FIELD_SENIORITY_LEVEL,
            years_experience_field=METADATA_CONFIG.FIELD_YEARS_EXPERIENCE,
            skill_name_field=METADATA_CONFIG.FIELD_SKILL_NAME,
            type_field=METADATA_CONFIG.FIELD_TYPE,
            type_skill=METADATA_CONFIG.TYPE_SKILL
        )
    
    async def execute(self, request: ChatRequestDto) -> ChatResult:
        parsed_query = self.query_parser.parse(request.question)
        
        query_embedding = self.embeddings_client.embed_query(request.question)
        metadata_filter = self._build_metadata_filter(request.filters, parsed_query)
        
        search_results = self.vector_store.search(
            query_embedding=query_embedding,
            limit=DEFAULT_LIMIT,
            filter_metadata=metadata_filter
        )
        
        if not search_results:
            return ChatResult(
                answer="No candidates found matching the specified criteria.",
                sources=[]
            )
        
        aggregated_candidates = self.candidate_aggregator.aggregate(search_results)
        filtered_candidates = self.filter_builder.filter_aggregated_candidates(
            aggregated_candidates,
            parsed_query
        )
        
        if not filtered_candidates:
            return ChatResult(
                answer="No candidates found matching the specified criteria.",
                sources=[]
            )
        
        ranked_candidates = self.candidate_ranker.rank(filtered_candidates, parsed_query)
        
        context = self._build_context_from_candidates(ranked_candidates)
        sources = self._extract_sources_from_candidates(ranked_candidates)
        
        human_prompt = self._get_human_prompt(context, request.question)
        system_prompt = self._get_system_prompt()
        
        llm_output = self.llm_client.generate_chat_completion(
            system_prompt=system_prompt,
            user_message=human_prompt,
            context=context
        )
        
        parsed_response = self._parse_and_validate_llm_output(llm_output)
        
        answer = self._format_final_answer(parsed_response)
        
        metadata = self._build_response_metadata(parsed_response)
        
        return ChatResult(
            answer=answer,
            sources=sources,
            metadata=metadata
        )
    
    def _build_metadata_filter(self, filters, parsed_query):
        conditions = []
        
        if filters:
            if filters.prepared is not None:
                conditions.append({PREPARED_KEY: filters.prepared})
                
            if filters.english_min:
                conditions.append({ENGLISH_LEVEL_NUM_MIN_KEY: ENGLISH_LEVEL_MAP.get(filters.english_min.upper(), 0)})
                
            if filters.candidate_ids:
                conditions.append({METADATA_CONFIG.FIELD_CANDIDATE_ID: {IN_OPERATOR: filters.candidate_ids}})
        
        query_filters = self.filter_builder.build_candidate_filters(parsed_query)
        conditions.extend(query_filters)
        
        tech_filter = self.filter_builder.build_technology_filters(parsed_query)
        if tech_filter:
            conditions.append(tech_filter)
        
        if not conditions:
            return None
        
        if len(conditions) == 1:
            return conditions[0]
        
        return {"$and": conditions}
    
    def _build_context_from_candidates(self, candidates: List) -> str:
        context_parts = []
        for candidate in candidates:
            for document in candidate.documents:
                context_parts.append(document)
        return CONTEXT_SEPARATOR.join(context_parts)
    
    def _extract_sources_from_candidates(self, candidates: List) -> List[ChatSource]:
        sources = []
        for candidate in candidates:
            for i, document in enumerate(candidate.documents):
                sources.append(ChatSource(
                    candidate_id=candidate.candidate_id,
                    section=candidate.metadata.get(METADATA_CONFIG.FIELD_TYPE, UNKNOWN_VALUE),
                    content=document[:DEFAULT_CONTENT_LIMIT] + CONTENT_SUFFIX if len(document) > DEFAULT_CONTENT_LIMIT else document,
                    score=candidate.all_scores[i] if i < len(candidate.all_scores) else 0.0
                ))
        return sources
    
    def _build_context(self, search_results: List) -> str:
        context_parts = []
        for document, metadata, score in search_results:
            context_parts.append(document)
        return CONTEXT_SEPARATOR.join(context_parts)
    
    def _extract_sources(self, search_results: List) -> List[ChatSource]:
        sources = []
        for document, metadata, score in search_results:
            sources.append(ChatSource(
                candidate_id=metadata.get(METADATA_CONFIG.FIELD_CANDIDATE_ID, UNKNOWN_VALUE),
                section=metadata.get(METADATA_CONFIG.FIELD_TYPE, UNKNOWN_VALUE),
                content=document[:DEFAULT_CONTENT_LIMIT] + CONTENT_SUFFIX if len(document) > DEFAULT_CONTENT_LIMIT else document,
                score=score
            ))
        return sources
    
    def _get_system_prompt(self) -> str:
        return load_prompt(CHAT_SYSTEM_FILE)
    
    def _get_human_prompt(self, context: str, question: str) -> str:
        human_template = load_prompt(CHAT_HUMAN_FILE)
        return human_template.format(context=context, input=question)
    
    def _parse_and_validate_llm_output(self, llm_output: str) -> LlmResponseSchema:
        try:
            json_start = llm_output.find('{')
            json_end = llm_output.rfind('}') + 1
            
            if json_start == -1 or json_end == 0:
                raise ValueError("No JSON object found in LLM output")
            
            json_str = llm_output[json_start:json_end]
            data = json.loads(json_str)
            
            if "justification" not in data:
                raise ValueError("Missing required field: justification")
            
            selected_candidate = None
            if data.get("selected_candidate") is not None:
                candidate_data = data["selected_candidate"]
                
                if "fullname" not in candidate_data or "candidate_id" not in candidate_data or "rank" not in candidate_data:
                    raise ValueError("selected_candidate must contain fullname, candidate_id, and rank")
                
                selected_candidate = SelectedCandidate(
                    fullname=candidate_data["fullname"],
                    candidate_id=candidate_data["candidate_id"],
                    rank=candidate_data["rank"]
                )
            
            return LlmResponseSchema(
                selected_candidate=selected_candidate,
                justification=data["justification"]
            )
        
        except (json.JSONDecodeError, ValueError, KeyError) as e:
            raise ValueError(f"Failed to parse LLM output as valid JSON schema: {str(e)}")
    
    def _format_final_answer(self, parsed_response: LlmResponseSchema) -> str:
        if parsed_response.selected_candidate is None:
            return f"No candidate selected.\n\n{parsed_response.justification}"
        
        return (
            f"Selected Candidate: {parsed_response.selected_candidate.fullname} "
            f"(ID: {parsed_response.selected_candidate.candidate_id}, Rank: {parsed_response.selected_candidate.rank})\n\n"
            f"Justification: {parsed_response.justification}"
        )
    
    def _build_response_metadata(self, parsed_response: LlmResponseSchema) -> dict:
        metadata = {
            "justification": parsed_response.justification
        }
        
        if parsed_response.selected_candidate is not None:
            metadata["selected_candidate"] = {
                "fullname": parsed_response.selected_candidate.fullname,
                "candidate_id": parsed_response.selected_candidate.candidate_id,
                "rank": parsed_response.selected_candidate.rank
            }
        else:
            metadata["selected_candidate"] = None
        
        return metadata

File: src/dotnet/Core/Application/UseCases/AskQuestionUseCase.cs
using Rag.Candidates.Core.Application.DTOs;
using Rag.Candidates.Core.Application.Interfaces;
using Rag.Candidates.Core.Application.Services;
using Rag.Candidates.Core.Domain.Configuration;
using Rag.Candidates.Core.Domain.Entities;

namespace Rag.Candidates.Core.Application.UseCases;

public sealed class AskQuestionUseCase
{
    private readonly IEmbeddingsClient _embeddingsClient;
    private readonly IVectorStore _vectorStore;
    private readonly IStructuredLlmClient<LlmJustificationSchema> _structuredLlmClient;
    private readonly IResourceLoader _resourceLoader;
    private readonly string _collection;
    private readonly VectorMetadataConfig _metadataConfig;
    private readonly QueryParser _queryParser;
    private readonly CandidateAggregator _candidateAggregator;
    private readonly CandidateRanker _candidateRanker;
    private readonly MetadataFilterBuilder _filterBuilder;

    private const int DefaultLimit = 6;
    private const string InOperator = "$in";
    private const string AndOperator = "$and";
    private const string GteOperator = "$gte";
    private const string ContextSeparator = "\n\n";
    private const string NewLine = "\n";
    private const string UnknownValue = "unknown";
    private const int DefaultContentLimit = 200;
    private const string ContentSuffix = "...";
    private const string ChatSystemFile = "chat_system.md";
    private const string ChatHumanFile = "chat_human.md";
    private const string NoCandidatesFoundMessage = "No candidates found matching the specified criteria.";
    private const string PreparedKey = "prepared";
    private const string EnglishLevelNumMinKey = "english_level_num_min";

    public AskQuestionUseCase(
        IEmbeddingsClient embeddingsClient,
        IVectorStore vectorStore,
        IStructuredLlmClient<LlmJustificationSchema> structuredLlmClient,
        IResourceLoader resourceLoader,
        VectorStorageSettings vectorSettings)
    {
        _embeddingsClient = embeddingsClient;
        _vectorStore = vectorStore;
        _structuredLlmClient = structuredLlmClient;
        _resourceLoader = resourceLoader;
        _collection = vectorSettings.CollectionName;
        
        _metadataConfig = VectorMetadataConfig.Default;
        _queryParser = new QueryParser();
        _candidateAggregator = new CandidateAggregator(_metadataConfig.FieldCandidateId);
        _candidateRanker = new CandidateRanker(RankingWeights.Default());
        _filterBuilder = new MetadataFilterBuilder(
            candidateIdField: _metadataConfig.FieldCandidateId,
            seniorityField: _metadataConfig.FieldSeniorityLevel,
            yearsExperienceField: _metadataConfig.FieldYearsExperience,
            skillNameField: _metadataConfig.FieldSkillName,
            typeField: _metadataConfig.FieldType,
            typeSkill: _metadataConfig.TypeSkill
        );
    }

    public async Task<ChatResult> ExecuteAsync(ChatRequestDto request, CancellationToken ct = default)
    {
        var parsedQuery = _queryParser.Parse(request.Question);
        
        var queryEmbedding = await _embeddingsClient.EmbedAsync([request.Question], ct);
        var metadataFilter = BuildMetadataFilter(request.Filters, parsedQuery);

        var searchResults = await _vectorStore.SearchAsync(
            collection: _collection,
            queryVector: queryEmbedding[0],
            limit: DefaultLimit,
            filter: metadataFilter,
            ct: ct
        );

        if (searchResults.Length == 0)
        {
            return new ChatResult
            {
                Answer = NoCandidatesFoundMessage,
                Sources = Array.Empty<ChatSource>()
            };
        }

        var searchResultsList = searchResults.Select(r => (r.Document, r.Metadata, (double)r.Score)).ToList();
        var aggregatedCandidates = _candidateAggregator.Aggregate(searchResultsList);
        var filteredCandidates = _filterBuilder.FilterAggregatedCandidates(aggregatedCandidates, parsedQuery);

        if (filteredCandidates.Count == 0)
        {
            return new ChatResult
            {
                Answer = NoCandidatesFoundMessage,
                Sources = Array.Empty<ChatSource>()
            };
        }

        var rankedCandidates = _candidateRanker.Rank(filteredCandidates, parsedQuery);

        var context = BuildContextFromCandidates(rankedCandidates);
        var sources = ExtractSourcesFromCandidates(rankedCandidates);

        var systemPrompt = await GetSystemPrompt(ct);
        var humanPrompt = await GetHumanPrompt(context, request.Question, ct);

        var chatContext = new ChatContext(systemPrompt, humanPrompt, context);
        var llmJustification = await _structuredLlmClient.GenerateStructuredAsync(chatContext, ct);
        
        var parsedResponse = BuildFinalResponse(llmJustification, rankedCandidates);
        
        var answer = FormatFinalAnswer(parsedResponse);
        
        var metadata = BuildResponseMetadata(parsedResponse);

        return new ChatResult
        {
            Answer = answer,
            Sources = sources,
            Metadata = metadata
        };
    }

    private Dictionary<string, object>? BuildMetadataFilter(ChatFilters? filters, Domain.Entities.ParsedQuery parsedQuery)
    {
        var techFilter = _filterBuilder.BuildTechnologyFilters(parsedQuery);
        if (techFilter != null)
        {
            return techFilter;
        }

        var conditions = new List<Dictionary<string, object>>();

        if (filters != null)
        {
            if (filters.Prepared.HasValue)
            {
                conditions.Add(new Dictionary<string, object> { [PreparedKey] = filters.Prepared.Value });
            }

            if (!string.IsNullOrEmpty(filters.EnglishMin))
            {
                var englishLevelNum = EnglishLevelToNum(filters.EnglishMin);
                conditions.Add(new Dictionary<string, object> { [EnglishLevelNumMinKey] = new Dictionary<string, object> { [GteOperator] = englishLevelNum } });
            }

            if (filters.CandidateIds?.Length > 0)
            {
                conditions.Add(new Dictionary<string, object> { [_metadataConfig.FieldCandidateId] = new Dictionary<string, object> { [InOperator] = filters.CandidateIds } });
            }
        }

        var queryFilters = _filterBuilder.BuildCandidateFilters(parsedQuery);
        conditions.AddRange(queryFilters);

        if (conditions.Count == 0)
            return null;

        if (conditions.Count == 1)
            return conditions[0];

        return new Dictionary<string, object> { [AndOperator] = conditions };
    }

    private string BuildContextFromCandidates(List<RankedCandidate> candidates)
    {
        var contextParts = new List<string>();
        for (int i = 0; i < candidates.Count; i++)
        {
            var candidate = candidates[i];
            var rank = i + 1;
            
            var fullname = candidate.Metadata.TryGetValue(_metadataConfig.FieldFullname, out var fn) 
                ? fn.ToString() ?? candidate.CandidateId 
                : candidate.CandidateId;
            var header = $"=== CANDIDATE #{rank}: {fullname} (ID: {candidate.CandidateId}) ===";
            contextParts.Add(header);
            
            contextParts.AddRange(candidate.Documents);
            
            contextParts.Add("");
        }
        
        return string.Join(ContextSeparator, contextParts);
    }

    private ChatSource[] ExtractSourcesFromCandidates(List<RankedCandidate> candidates)
    {
        var sources = new List<ChatSource>();
        foreach (var candidate in candidates)
        {
            for (int i = 0; i < candidate.Documents.Count; i++)
            {
                var document = candidate.Documents[i];
                var score = i < candidate.AllScores.Count ? candidate.AllScores[i] : 0.0;
                var section = candidate.Metadata.TryGetValue(_metadataConfig.FieldType, out var type) 
                    ? type.ToString() ?? UnknownValue 
                    : UnknownValue;
                var content = document.Length > DefaultContentLimit 
                    ? document[..DefaultContentLimit] + ContentSuffix 
                    : document;

                sources.Add(new ChatSource
                {
                    CandidateId = candidate.CandidateId,
                    Section = section,
                    Content = content,
                    Score = (float)score
                });
            }
        }
        return sources.ToArray();
    }

    private static string BuildContext((string Document, Dictionary<string, object> Metadata, float Score)[] searchResults)
    {
        var contextParts = searchResults.Select(result => result.Document);
        return string.Join(ContextSeparator, contextParts);
    }

    private ChatSource[] ExtractSources((string Document, Dictionary<string, object> Metadata, float Score)[] searchResults)
    {
        return searchResults.Select(result =>
        {
            var candidateId = result.Metadata.TryGetValue(_metadataConfig.FieldCandidateId, out var id) ? id.ToString() ?? UnknownValue : UnknownValue;
            var section = result.Metadata.TryGetValue(_metadataConfig.FieldType, out var type) ? type.ToString() ?? UnknownValue : UnknownValue;
            var content = result.Document.Length > DefaultContentLimit 
                ? result.Document[..DefaultContentLimit] + ContentSuffix 
                : result.Document;

            return new ChatSource
            {
                CandidateId = candidateId,
                Section = section,
                Content = content,
                Score = result.Score
            };
        }).ToArray();
    }

    private async Task<string> GetSystemPrompt(CancellationToken ct)
    {
        return await _resourceLoader.LoadPromptAsync(ChatSystemFile, ct);
    }

    private async Task<string> GetHumanPrompt(string context, string question, CancellationToken ct)
    {
        var humanTemplate = await _resourceLoader.LoadPromptAsync(ChatHumanFile, ct);
        return humanTemplate.Replace("{context}", context).Replace("{input}", question);
    }

    private static int EnglishLevelToNum(string level)
    {
        if (string.IsNullOrEmpty(level))
            return 0;

        return level.ToUpperInvariant() switch
        {
            "A1" => 1,
            "A2" => 2,
            "B1" => 3,
            "B2" => 4,
            "C1" => 5,
            "C2" => 6,
            "BASIC" => 2,
            "CONVERSATIONAL" => 3,
            "FLUENT" => 5,
            "NATIVE" => 6,
            "ADVANCED" => 5,
            _ => 0
        };
    }
    
    private LlmResponseSchema BuildFinalResponse(LlmJustificationSchema llmJustification, List<RankedCandidate> rankedCandidates)
    {
        if (rankedCandidates.Count == 0)
        {
            return new LlmResponseSchema
            {
                SelectedCandidate = null,
                Justification = llmJustification.Justification
            };
        }
        
        var topCandidate = rankedCandidates[0];
        var fullname = topCandidate.Metadata.TryGetValue(_metadataConfig.FieldFullname, out var fn) 
            ? fn.ToString() ?? topCandidate.CandidateId 
            : topCandidate.CandidateId;
        
        var selectedCandidate = new SelectedCandidate
        {
            Fullname = fullname,
            CandidateId = topCandidate.CandidateId,
            Rank = 1
        };
        
        return new LlmResponseSchema
        {
            SelectedCandidate = selectedCandidate,
            Justification = llmJustification.Justification
        };
    }
    
    private string FormatFinalAnswer(LlmResponseSchema parsedResponse)
    {
        if (parsedResponse.SelectedCandidate is null)
        {
            return $"No candidate selected.\n\n{parsedResponse.Justification}";
        }
        
        return $"Selected Candidate: {parsedResponse.SelectedCandidate.Fullname} " +
               $"(ID: {parsedResponse.SelectedCandidate.CandidateId}, Rank: {parsedResponse.SelectedCandidate.Rank})\n\n" +
               $"Justification: {parsedResponse.Justification}";
    }
    
    private Dictionary<string, object> BuildResponseMetadata(LlmResponseSchema parsedResponse)
    {
        var metadata = new Dictionary<string, object>
        {
            ["justification"] = parsedResponse.Justification
        };
        
        if (parsedResponse.SelectedCandidate is not null)
        {
            metadata["selected_candidate"] = new Dictionary<string, object>
            {
                ["fullname"] = parsedResponse.SelectedCandidate.Fullname,
                ["candidate_id"] = parsedResponse.SelectedCandidate.CandidateId,
                ["rank"] = parsedResponse.SelectedCandidate.Rank
            };
        }
        else
        {
            metadata["selected_candidate"] = null!;
        }
        
        return metadata;
    }
}